tests/ragged_inference/test_ragged_qk_dotprod.py

```diff
@@ -0,0 +1,135 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+import math
+
+import pytest
+import torch
+from xformers.helpers.test_utils import bf16_cuda
+from xformers.triton.garbage_pad_ragged_acts import (
+    RaggedActivations,
+    get_acts_offset_per_seq,
+)
+from xformers.triton.k_ragged_qk_dotprod import mem_efficient_fw
+from xformers.triton.utils import gpu_capabilities_older_than_70
+
+
+def _make_seq(n_ctx: int, value: int, d_model: int):
+    return torch.full([n_ctx, d_model], value, **bf16_cuda())
+
+
+_triton_available = True
+# except ImportError as e:
+#     err = f"Triton is not available, some optimizations will "f"not be tested.{e}"
+#     logging.warning(
+#         err
+#     )
+#     print(err)
+#     _triton_available = False
+
+
+# Testing odd shapes on purpose
+# Shapes correspond to (B, M&N, L)
+# TODO: test with M & N different from each other
+SHAPES = [
+    (1, 384, 128),
+    (8, 384, 128),
+    (8, 784, 512),
+    (16, 1024, 1024),
+    (2, 2048, 384),
+    # (1, 18, 128), # FIXME
+]
+
+
+def attention_pytorch(q, k):
+    # attention matrix
+    q = q / math.sqrt(q.size(-1))
+    a = q @ k.transpose(-2, -1)
+    return a
+
+
+@pytest.mark.skipif(not _triton_available, reason="Triton is not available")
+@pytest.mark.skipif(
+    not _triton_available or gpu_capabilities_older_than_70(),
+    reason="Triton requires a SM70+ GPU",
+)
+@pytest.mark.parametrize("shape", SHAPES)
+@pytest.mark.parametrize("dtype", [torch.float32])
+def test_mem_efficient_attention_parity(shape, dtype):
+    q = torch.rand(shape, dtype=dtype, device=torch.device("cuda"))
+    k = torch.rand(shape, dtype=dtype, device=torch.device("cuda"))
+
+    res_pytorch = attention_pytorch(q, k)
+    res_me = mem_efficient_fw(q=q, k=k)
+    # print(res_me)
+    assert torch.mean(torch.abs(res_pytorch - res_me)) < 0.2
```

Comment:
@nottombrown so this works not with a catch a little more stringent ? (toch.max()) torch.allclose() would probably be a good way to really test that